src/lib/support/tests/TestCHIPMemString.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements a unit test suite for CHIP Memory Management
+ *      code functionality.
+ *
+ */
+
+#include <inttypes.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <lib/support/CHIPMemString.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/ScopedBuffer.h>
+#include <lib/support/Span.h>
+#include <lib/support/UnitTestRegistration.h>
+#include <nlunit-test.h>
+
+using namespace chip;
+using namespace chip::Platform;
+
+// =================================
+//      Unit tests
+// =================================
+
+namespace {
+template <size_t testStrLen>
+struct TestBuffers
+{
+    char correctSizeBuf[testStrLen + 1];
+    char tooSmallBuf[testStrLen];
+    char wayTooSmallBuf[1];
+    char tooBigBuf[testStrLen + 10];
+    void Reset()
+    {
+        memset(correctSizeBuf, 0, sizeof(correctSizeBuf));
+        memset(tooSmallBuf, 0, sizeof(tooSmallBuf));
+        memset(wayTooSmallBuf, 0, sizeof(wayTooSmallBuf));
+        memset(tooBigBuf, 0, sizeof(tooBigBuf));
+    }
+    void CheckCorrectness(nlTestSuite * inSuite, const char * testStr)
+    {
+        // correctSizeBuf and tooBigBuf should have the complete string.
+        NL_TEST_ASSERT(inSuite, correctSizeBuf[testStrLen] == '\0');
+        NL_TEST_ASSERT(inSuite, tooBigBuf[testStrLen] == '\0');
+        NL_TEST_ASSERT(inSuite, strcmp(correctSizeBuf, testStr) == 0);
+        NL_TEST_ASSERT(inSuite, strcmp(tooBigBuf, testStr) == 0);
+
+        // wayTooSmallBuf is tiny and thus will only have the null terminator
+        NL_TEST_ASSERT(inSuite, wayTooSmallBuf[0] == '\0');
+
+        // tooSmallBuf should still have a null terminator on the end
+        NL_TEST_ASSERT(inSuite, tooSmallBuf[testStrLen - 1] == '\0');
+    }
+};
+
+} // namespace
+
+static void TestCopyString(nlTestSuite * inSuite, void * inContext)
+{
+    constexpr char testWord[] = "testytest";
+    ByteSpan testWordSpan     = ByteSpan(reinterpret_cast<const uint8_t *>(testWord), sizeof(testWord));
```

Comment:
```suggestion
    ByteSpan testWordSpan     = ByteSpan(reinterpret_cast<const uint8_t *>(testWord), sizeof(testWord) - 1);
```
Because the ByteSpan should be just the text, not counting the null-terminator...